lambdas/address/src/main/java/uk/gov/di/ipv/cri/address/api/handler/AddressDateLinker.java

```diff
@@ -0,0 +1,102 @@
+            "Too many Addresses for AddressDateLinker to process.";
+
+    public static final String ERROR_COULD_NOT_DETERMINE_CURRENT_ADDRESS =
+            "Two addresses found but could not determine which address is a CURRENT addressType.";
+
+    public static final String ERROR_ADDRESS_LINKER_NOT_NEEDED =
+            "Address Linker found PREVIOUS address but validUntil was already set - automatic date linking failed.";
+
+    @ExcludeFromGeneratedCoverageReport
+    private AddressDateLinker() {
+        throw new IllegalStateException("Static Utility class");
+    }
+
+    public static void linkCurrentAndPreviousAddressesDates(List<CanonicalAddress> addresses)
+            throws AddressProcessingException {
+
+        switch (addresses.size()) {
+            case 0:
+                throw new AddressProcessingException(ERROR_NO_ADDRESSES);
+            case 1:
+                if (isNotCurrentAddress(addresses.get(0))) {
+                    throw new AddressProcessingException(ERROR_SINGLE_ADDRESS_NOT_CURRENT);
+                }
+                return;
+            case 2:
+                processAddresses(addresses);
+                return;
+            default:
+                // We cannot link multiple PREVIOUS address dates as they are null.
+                // Only the CURRENT address has date information.
+                throw new AddressProcessingException(ERROR_TOO_MANY_ADDRESSES);
+        }
+    }
+
+    private static void processAddresses(List<CanonicalAddress> addresses)
+            throws AddressProcessingException {
+
+        CanonicalAddress currentAddress = null;
+        CanonicalAddress previousAddress = null;
+
+        CanonicalAddress address0 = addresses.get(0);
+        CanonicalAddress address1 = addresses.get(1);
+
+        if (isCurrentAddress(address0) && isNotCurrentAddress(address1)) {
+            currentAddress = address0;
+            previousAddress = address1;
+        } else if (isCurrentAddress(address1) && isNotCurrentAddress(address0)) {
+            currentAddress = address1;
+            previousAddress = address0;
+        } else if (isCurrentAddress(address0) && isCurrentAddress(address1)) {
+
+            // This an edge case where there are two CURRENT address.
+            // Date linking is not performed.
+
+            return;
+        } else {
+            throw new AddressProcessingException(ERROR_COULD_NOT_DETERMINE_CURRENT_ADDRESS);
+        }
+
+        // When AddressCRI Front is updated to set validUntil the linker needs to be removed
+        // else it would trample the dates already set
+        if (previousAddress.getValidUntil() != null) {
+
+            // Not safe to automatically process dates in addresses
+            throw new AddressProcessingException(ERROR_ADDRESS_LINKER_NOT_NEEDED);
+        }
+
+        // At this point the current address is known
+        LocalDate validFrom = currentAddress.getValidFrom();
```

Comment:
This line can be removed and the line beneath can be updated as follows:
`previousAddress.setValidUntil(currentAddress.getValidFrom());